flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/calls/ScalarOperatorGens.scala

```diff
@@ -1767,6 +1767,56 @@ object ScalarOperatorGens {
     }
   }
 
+  /**
+   * Return null when at least one of the elements is null.
+   */
+  def generateGreatestLeast(
+      resultType: LogicalType,
+      elements: Seq[GeneratedExpression],
+      greatest: Boolean = true)
+  : GeneratedExpression = {
+    val Seq(result, cur, nullTerm) = newNames("result", "cur", "nullTerm")
+    val widerType = toScala(findCommonType(elements.map(element => element.resultType)))
+      .orElse(throw new CodeGenException(s"Unable to find common type for $elements."))
+    val resultTypeTerm = boxedTypeTermForType(widerType.get)
+
+    def castIfNumeric(t: GeneratedExpression): String = {
+      if (TypeCheckUtils.isNumeric(widerType.get)) {
+         s"${numericCasting(t.resultType, widerType.get).apply(t.resultTerm)}"
+      } else {
+         s"${t.resultTerm}"
+      }
+    }
+
+    val elementsCode = elements.map { element =>
+      s"""
+         | ${element.code}
+         | if (!$nullTerm) {
+         |   $resultTypeTerm $cur = ${castIfNumeric(element)};
+         |   if (${element.nullTerm}) {
+         |     $nullTerm = true;
+         |   } else {
+         |     int compareResult = $result.compareTo($cur);
+         |     if ($greatest && compareResult < 0 || compareResult > 0 && !$greatest) {
```

Comment:
nit: this is generated code, so it's semi-important, but I'd find this a bit more readable:

```
if (($greatest && compareResult < 0) || (!$greatest && compareResult > 0)) {
```